File: src/main/java/com/dev/photoCatalog/service/PhotoshootService.java
package com.dev.photoCatalog.service;

import com.dev.photoCatalog.dto.PhotoshootDTO;
import com.dev.photoCatalog.model.Photoshoot;
import com.dev.photoCatalog.model.Photo;
import com.dev.photoCatalog.model.Location;
import com.dev.photoCatalog.repository.PhotoshootRepository;
import com.dev.photoCatalog.repository.PhotoRepository;
import com.dev.photoCatalog.repository.LocationRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.jdbc.core.JdbcTemplate;
import jakarta.transaction.Transactional;

import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
public class PhotoshootService {

    @Autowired
    private PhotoshootRepository photoshootRepository;

    @Autowired
    private PhotoRepository photoRepository;

    @Autowired
    private LocationRepository locationRepository;

    @Autowired
    private JdbcTemplate jdbcTemplate;

    public PhotoshootDTO convertToDTO(Photoshoot photoshoot) {
        return new PhotoshootDTO(
                photoshoot.getPhotoshootID(),
                photoshoot.getDate(),
                photoshoot.getLocationID()  // Use the locationID directly
        );
    }

    public Photoshoot convertToEntity(PhotoshootDTO dto) {
        Photoshoot photoshoot = new Photoshoot();
        photoshoot.setPhotoshootID(dto.getPhotoshootID());
        photoshoot.setDate(dto.getDate());
        photoshoot.setLocationID(dto.getLocationID());  // Use locationID directly
        return photoshoot;
    }

    // Get all photoshoots and return as DTOs
    public List<PhotoshootDTO> getAllPhotoshoots() {
        return photoshootRepository.findAll().stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    // Get a specific photoshoot by ID
    public Optional<PhotoshootDTO> getPhotoshootById(int id) {
        return photoshootRepository.findById(id).map(this::convertToDTO);
    }

    // Add a new photoshoot
    public PhotoshootDTO addPhotoshoot(PhotoshootDTO dto) {
        Photoshoot savedPhotoshoot = photoshootRepository.save(convertToEntity(dto));
        return convertToDTO(savedPhotoshoot);
    }

    // Update an existing photoshoot
    public Optional<PhotoshootDTO> updatePhotoshoot(int id, PhotoshootDTO dto) {
        return photoshootRepository.findById(id).map(existing -> {
            existing.setDate(dto.getDate());

            // Use locationID directly instead of fetching the Location object
            existing.setLocationID(dto.getLocationID());

            Photoshoot updated = photoshootRepository.save(existing);
            return convertToDTO(updated);
        });
    }

    // Delete a photoshoot by ID
    public void deletePhotoshoot(int id) {
        photoshootRepository.deleteById(id);
    }

    // Get all photos associated with a specific photoshoot by querying the junction table
    public List<Photo> getAllPhotosForPhotoshoot(int photoshootId) {
        return photoRepository.findAllPhotosForPhotoshoot(photoshootId);
    }

    // Get all locations for a specific photoshoot by querying the junction table
    public List<Location> getAllLocationsForPhotoshoot(int photoshootId) {
        return locationRepository.findAllLocationsForPhotoshoot(photoshootId);
    }

    // Add a photo to a photoshoot using photoGUID
    @Transactional
    public void addPhotoToPhotoshoot(int photoshootID, UUID photoGUID) {
        Photoshoot photoshoot = photoshootRepository.findById(photoshootID)
                .orElseThrow(() -> new IllegalArgumentException("Photoshoot not found"));

        Photo photo = photoRepository.findByPhotoGUID(photoGUID.toString())
                .orElseThrow(() -> new IllegalArgumentException("Photo not found"));

        String sql = "INSERT INTO PhotoshootPhotoJunction (photoshootID, photoGUID) VALUES (?, ?)";
        jdbcTemplate.update(sql, photoshoot.getPhotoshootID(), photo.getPhotoGUID());
    }

    // Check if a photoshoot exists by ID
    public boolean photoshootExists(int id) {
        return photoshootRepository.existsById(id);
    }
}

File: frontend/src/photoshootService.js
import axios from 'axios';

//get all photoshoots.
export async function getAllPhotoshoots() {
  try {
    const response = await axios.get("http://localhost:8080/photoshoot");
    if (response.status !== 200) {
      console.log(`${response.status} - ${response.statusText}`);
    }
    //console.log("hit check");
    return response.data;
  } catch (error) {
    console.log("Error fetching photoshoots:", error);
  }
}

//get a specific photoshoot by ID.
export async function getPhotoshootById(id) {
  try {
    const response = await axios.get(`http://localhost:8080/photoshoot/${id}`);
    if (response.status !== 200) {
      console.log(`${response.status} - ${response.statusText}`);
    }
    return response.data;
  } catch (error) {
    console.log("Error fetching photoshoot:", error);
  }
}

//make a new photoshoot.
export async function addPhotoshoot(date) {
    try {
      const response = await axios.post("http://localhost:8080/photoshoot", {
        //keeping a placeholder locationID for now. Will come back and allow users to select location later on
        locationID: 1, 
        date: date,
      });
      if (response.status !== 200) {
        console.log(`${response.status} - ${response.statusText}`);
      }
      return response.data;
    } catch (error) {
      console.log("Error creating photoshoot:", error);
    }
  }
  

 //add a photo to a photoshoot using the photoGUID.
export async function addPhotoToPhotoshoot(photoshootID, photoGUID) {
  try {
    const response = await axios.post(`http://localhost:8080/photoshoot/${photoshootID}/addPhoto/${photoGUID}`);
    if (response.status !== 200) {
      console.log(`${response.status} - ${response.statusText}`);
    }
  } catch (error) {
    console.log("Error adding photo to photoshoot:", error);
  }
}

 //get all photos for a specific photoshoot.
export async function getAllPhotosForPhotoshoot(photoshootID) {
  try {
    const response = await axios.get(`http://localhost:8080/photoshoot/${photoshootID}/photo`);
    if (response.status !== 200) {
      console.log(`${response.status} - ${response.statusText}`);
    }
    return response.data;
  } catch (error) {
    console.log("Error grabbing photos for photoshoot:", error);
  }
}
